### PR TITLE
feat: initial work

### PR DESCRIPTION
**Pull request by tam5**

### CODE REVIEW

File: src/Services/TokenService.php [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+     *
+     * @return void
+     */
+    public function __construct(array $options = [])
+    {
+        $this->baseUrl = $options['base_url'];
+        $this->clientId = $options['client_id'];
+        $this->clientSecret = $options['client_secret'];
+
+        $this->http = new Http();
```
Comment (tam5) [inline src/Services/TokenService.php]: make this injectable

File: src/Services/ServiceFactory.php [hunk 2]
```diff
@@ -0,0 +1,8 @@
+<?php
+
+namespace LiveIntent\Services;
+
+class ServiceFactory
+{
+    //
+}
```
Comment (tam5) [inline src/Services/ServiceFactory.php]: not used, remove

File: src/Services/TokenService.php [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+        $payload = $response->throw()->json();
+
+        $this->accessToken = $payload['access_token'];
+        $this->tokenType = $payload['token_type'];
+        $this->expiresAt = Carbon::now()->addSeconds($payload['expires_in'] - $this->bufferSeconds);
+    }
+
+    /**
+     * Check if a new tokens should be generated.
```
Comment (tam5) [inline src/Services/TokenService.php]: english

File: src/Support/HasAttributes.php [hunk 4]
```diff
@@ -0,0 +1,1039 @@
+<?php
+
```
Comment (tam5) [inline src/Support/HasAttributes.php:2]: i guess we should credit this file

File: tests/ExampleTest.php [hunk 5]
```diff
@@ -2,10 +2,19 @@
...
+
 class ExampleTest extends TestCase
 {
-    public function testTrueIsTrue()
+    public function testExampleStuff()
     {
-        $this->assertTrue(true);
+        $li = new LiveIntentClient([
+            'client_id' => 'ari',
+            'client_secret' => '93f129a60f17264feab81a260256f13e',
```
Comment (tam5) [inline tests/ExampleTest.php]: envs / stubs

File: tests/Services/LineItemServiceTest.php [hunk 6]
```diff
@@ -0,0 +1,96 @@
+<?php
+
+namespace Tests\Services;
+
+use LiveIntent\LineItem;
+use LiveIntent\Services\LineItemService;
+use LiveIntent\Exceptions\InvalidRequestException;
+use LiveIntent\Exceptions\InvalidArgumentException;
+
+class LineItemServiceTest extends ServiceTestCase
```
Comment (tam5) [inline tests/Services/LineItemServiceTest.php:10]: refactor

File: src/Client/BaseClient.php [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+     */
+    public function request($method, $path, $data = null, $opts = [])
+    {
+        return $this
+            ->newPendingRequest()
+            ->baseUrl($this->baseUrl)
+            ->withToken($this->tokenService->token(), $this->tokenService->tokenType())
+            ->withBody(json_encode($data), 'application/json')
+            ->acceptJson()
+            ->retry($opts['tries'] ?? $this->tries, $opts['retryDelay'] ?? $this->retryDelay)
```
Comment (tam5) [inline src/Client/BaseClient.php:116]: let's add timeout config explicitly

File: src/Client/BaseClient.php [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+     * @var \LiveIntent\Services\TokenService
+     */
+    private $tokenService;
+
+    /**
+     * Create a new instance.
+     *
+     * @return void
+     */
+    public function __construct(array $options = [])
```
Comment (tam5) [inline src/Client/BaseClient.php:82]: read tries and delay from options

File: src/Services/AbstractService.php [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+     * @return \LiveIntent\Resource
+     */
+    public function update($attributes, $opts = [])
+    {
+        $payload = (array) $attributes;
+        $id = $payload['id'] ?? null;
+
+        if ($attributes instanceof Resource) {
+            $id = $attributes->id;
+            $payload = array_merge($attributes->getDirty(), ['version' => $attributes->version]);
```
Comment (tam5) [inline src/Services/AbstractService.php:85]: not everyone will have a version

**Comment by tam5** [comment]
Should we allow cookie based auth? Or should we imply that loginAs would be proper usage?

File: src/Client/BaseClient.php [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+use Illuminate\Http\Client\Factory as IlluminateClient;
+
+class BaseClient extends IlluminateClient implements ClientInterface
+{
+    /**
+     * The base url for all api requests issued by this client.
+     *
+     * @var string
+     */
+    private $baseUrl = 'http://localhost:33001'; // TODO change
```
Comment (rrmanalo623) [inline src/Client/BaseClient.php]: I assume this will be from some env var or something, hence your TODO.

File: src/Services/TokenService.php [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+     *
+     * @return bool
+     */
+    private function needsNewTokens()
```
Comment (rrmanalo623) [inline src/Services/TokenService.php:152]: Is this plural or singular?

Comment (tam5) [reply src/Services/TokenService.php:152]: It's plural on purpose.

One day we'll add other grant types which may include refresh tokens or open id tokens, etc.

File: src/Services/AbstractService.php (hunk 9 above)
Comment (tam5) [reply src/Services/AbstractService.php:85]: it's ok for now

File: tests/ExampleTest.php (hunk 5 above)
Comment (hallas) [reply tests/ExampleTest.php]: I would recommend showing how to read these from elsewhere, e.g. environment variables instead of the static example.

**Review by hallas — CHANGES_REQUESTED** [review]
Left one comment. Thank you.

File: src/Services/AbstractService.php [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+        }
+
+        if ($id === null) {
+            throw Exceptions\InvalidArgumentException::factory($payload, 'Unable to find `id` for update operation');
+        }
+
+        return $this->request('post', $this->resourceUrl($id), $payload, $opts);
+    }
+
+    // /**
```
Comment (jderman-liveintent) [inline src/Services/AbstractService.php:95]: Are these coming in the next update?

File: src/Services/TokenService.php [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+     * @var string
+     */
+    private $baseUrl;
+
+    /**
+     * The default number of times a request should be retried.
+     *
+     * @var int
+     */
+    private $tries = 1;
```
Comment (jderman-liveintent) [inline src/Services/TokenService.php:22]: should these be in the .env?

Comment (tam5) [reply src/Services/TokenService.php:22]: no, they are runtime configurable

**Pull request merged**